retrieval/target_provider.go

```diff
@@ -117,19 +161,77 @@ func (p *sdTargetProvider) Targets() ([]Target, error) {
 		if addr.Target[len(addr.Target)-1] == '.' {
 			addr.Target = addr.Target[:len(addr.Target)-1]
 		}
-		endpoint.Host = fmt.Sprintf("%s:%d", addr.Target, addr.Port)
-		t := NewTarget(endpoint.String(), p.job.ScrapeTimeout(), baseLabels)
+		tp.endpoint.Host = fmt.Sprintf("%s:%d", addr.Target, addr.Port)
+		t := NewTarget(tp.endpoint.String(), tp.deadline, tp.baseLabels)
 		targets = append(targets, t)
 	}
+	return targets, err
+}
 
-	p.targets = targets
+// etcdTargetProvider provides target information from the etcd key-value store
+// based on a provided config file.
+type etcdTargetProvider struct {
+	srv        config.ServiceDiscoveryConfig
+	endpoint   *url.URL
+	baseLabels clientmodel.LabelSet
+	deadline   time.Duration
+}
+
+// Targets implements the TargetProvider interface.
+func (tp *etcdTargetProvider) Targets() ([]Target, error) {
```

Comment:
It's just a KV store - there are a million ways to do it but covering the reasonable ones should be fine, right?